lib/http2/connection.c

```diff
@@ -1715,7 +1715,9 @@ static h2o_http2_conn_t *create_conn(h2o_context_t *ctx, h2o_hostconf_t **hosts,
         }},
     };
 
-    h2o_http2_conn_t *conn = (void *)h2o_create_connection(sizeof(*conn), ctx, hosts, connected_at, &callbacks);
+    void *_conn = h2o_mem_alloc_recycle(&h2o_mem_req_allocator);
+    h2o_http2_conn_t *conn = (void *)h2o_create_connection_in_place(_conn, ctx, hosts, connected_at, &callbacks);
+    H2O_BUILD_ASSERT(sizeof(*conn) < *h2o_mem_req_allocator.memsize);
```

Comment:
I haven't checked, but are we correctly doing the check here? `H2O_BUILD_ASSERT` turns into a no-op when the provided argument is not a constant expression (see https://github.com/h2o/h2o/pull/1041).

I suspect we need a macro that defines the size of the chunks being recycle, and compare against that value instead of a thread-local.

As an alternative, I would not mind passing `size` to `h2o_mem_alloc_recycle` and doing an `assert` there. That branch will never be taken in production, so it would be very lightweight.